lib/l10n_utils/__init__.py

```diff
@@ -144,56 +143,18 @@ def get_locale(request):
     return getattr(request, "locale", settings.LANGUAGE_CODE)
 
 
-def get_accept_languages(request):
-    """
-    Parse the user's Accept-Language HTTP header and return a list of languages
-    """
-    languages = []
-    pattern = re.compile(r"^([A-Za-z]{2,3})(?:-([A-Za-z]{2})(?:-[A-Za-z0-9]+)?)?$")
-
-    parsed = parse_accept_lang_header(request.META.get("HTTP_ACCEPT_LANGUAGE", ""))
-
-    for lang, priority in parsed:
-        m = pattern.match(lang)
-
-        if not m:
-            continue
-
-        lang = m.group(1).lower()
-
-        # Check if the shorter code is supported. This covers obsolete long
-        # codes like fr-FR (should match fr) or ja-JP (should match ja)
-        if m.group(2) and lang not in settings.PROD_LANGUAGES:
-            lang += "-" + m.group(2).upper()
-
-        if lang not in languages:
-            languages.append(lang)
-
-    return languages
-
-
 def get_best_translation(translations, accept_languages):
-    # Look for the user's Accept-Language HTTP header to find another
-    # locale we can offer
-    for lang in accept_languages:
+    ranked = parse_accept_lang_header(accept_languages)
+    for lang, _ in ranked:
         if lang in translations:
             return lang
+        pre = lang.split("-")[0]
+        if pre in translations:
+            return pre
+        if pre in settings.CANONICAL_LOCALES:
```

Comment:
I don't think we can use this check here. I think we'll need to construct a mapping of locales to URL locales like we do in `bedrock.base.urlresolvers._get_langauge_map()` but for only the languages in `translations`, then check that. I think like this it could result in a redirect loop by sending people to an unsupported language on the page and then doing that again when it gets back here.

Reply:
Yeah I thought it would be more complicated but had some spare time and made a quick attempt. The test failures tell me there's more to consider.